### ISSUE TITLE
where is the  torch.nn.NLLLoss ?

### ISSUE THREAD

**Issue by susht3** [pytorch/pytorch#2227]
i want to find how NLLLoss calcuate the loss, but i can't find its code.

# loss
def nll_loss(input, target, weight=None, size_average=True, ignore_index=-100):
    r"""The negative log likelihood loss.
    See :class:`~torch.nn.NLLLoss` for details.

where is `~torch.nn.NLLLoss`  ?

**Comment by bheinzerling** [comment]
`torch.nn.NLLLoss` is here:

https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/loss.py?hl=64#L64

Like many modules, this module is just a thin wrapper around the corresponding function, in this case `nll_loss`:

line 131: `return F.nll_loss(...)`

Tracing the F import, we find:

`https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L644`

This function, in turn, calls:

`return _functions.thnn.NLLLoss(...)`

`_functions` is here:
`https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/thnn/loss.py`

But this is kind of a dead end, as there is no Python code implementing `NLLLoss`.

Like many low-level functions, `NLLLoss` is implemented in C. The PyTorch C code is here:

https://github.com/pytorch/pytorch/blob/master/torch/lib

`torch.nn` implementations are in THNN, which leads to the actual implementation:

https://github.com/pytorch/pytorch/blob/master/torch/lib/THNN/generic/ClassNLLCriterion.c#L5

**Comment by tengerye** [comment]
@bheinzerling @Soltius63 Hi, sorry to interrupt. Links you provided are broken now, is it possible to update them please?

**Comment by maqboolkhan** [comment]
Updated link: https://github.com/pytorch/pytorch/blob/master/aten/src/THCUNN/ClassNLLCriterion.cu

**Comment by akwako** [comment]
Updated link: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Loss.cu#L73

**Comment by akwako** [comment]
Updated link: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Loss.cu#L149
